lib/rouge/lexers/batch.rb

```diff
@@ -0,0 +1,165 @@
+        getmac gpresult gpupdate help hostname
+        icacls ifmember inuse ipconfig kill
+        label lgpo lodctr logman logoff logtime
+        makecab mapisend mbsacli mem
+        mklink mode more mountvol moveuser msg mshta msiexec msinfo32 mstsc
+        nbtstat net net1 netdom netsh netstat nlsinfo nltest now nslookup
+        ntbackup ntdsutil ntoskrnl ntrights nvspbind
+        openfiles
+        pathping perms ping popd portqry powercfg pngout
+        pnputil print printbrm prncnfg prnmngr procdump
+        psexec psfile psgetsid psinfo pskill pslist
+        psloggedon psloglist pspasswd psping
+        psservice psshutdown pssuspend
+        qbasic qgrep qprocess query quser qwinsta
+        rasdial recover reg reg1 regdump regedt32
+        regsvr32 regini replace reset restore
+        rundll32 rmtshare robocopy route rpcping run runas
+        sc scandisk schtasks setspn setx sfc
+        share shellrunas shortcut shutdown
+        sigcheck sleep slmgr sort strings subinacl
+        subst sysmon systeminfo
+        takeown taskkill tasklist telnet tftp
+        timeout tlist touch tracerpt tracert tree tscon
+        tsdiscon tskill tttracer typeperf tzutil
+        undelete unformat
+        verifier vmconnect vssadmin
+        w32tm waitfor wbadmin wecutil wevtutil wget
+        where whoami windiff winrm winrs wmic wpeutil wpr wusa wuauclt
+        wscript
+        xcopy
+      ).join('|')
+
+      state :basic do
+        # Comments
+        rule %r/\b(REM)\b.*$/i, Comment
+        # Empty Labels
+        rule %r/^::.*$/, Comment
+
+        # Labels
+        rule %r/:[a-z]+/i, Keyword::Constant
+
+        # Devices
+        rule %r/@?\b(#{DEVICES})\s*\b/i, Keyword::Reserved
+        rule %r/@?\b(#{DEVICESNUM}[0-9])\s*\b/i, Keyword::Reserved
+
+        # Lang Keywords
+        rule %r/@?\b(#{KEYWORDS})\s*\b/i, Name::Builtin::Pseudo
+        # Internal Commands
+        rule %r/@?\b(#{INTERNAL})\s*\b/i, Name::Builtin::Pseudo
+        # External Commands
+        rule %r/\b(#{EXTERNAL})(\.exe|\.com)?\s*\b/i, Name::Builtin::Pseudo
+
+        # Arguments to commands
+        rule %r/\b(#{ATTRIBUTES})\s*\b/i, Name::Attribute
+        rule %r/([\/\-+][A-Z]+)\s*/i, Name::Attribute
+
+        # Variable Expansions
+        rule %r/[\%!]+([a-z_$@#]+)[\%!]+/i, Name::Variable
+        # For Variables
+        rule %r/(\%+~?[A-Z]+\d?)/i, Keyword::Constant
+
+        rule %r/\b(set)\b/i, Keyword::Declaration
+
+        # Operators
+        rule %r/[%!()=<>&|~$*@]/, Operator
+
+      end
+
+      state :single_quotes do
+        rule %r/[']/, Str::Single, :pop!
+        rule %r/[^']+/, Text
+      end
+
+      state :double_quotes do
+        rule %r/["]/, Str::Double, :pop!
+        rule %r/[^"]+/, Text
+      end
+
+      state :backtick do
+        rule %r/[`]/, Str::Backtick, :pop!
+        rule %r/[^`]+/, Text
+      end
+
+      state :data do
+        rule %r/\s+/, Text
+        rule %r/\^/, Str::Escape
```

Comment:
Shouldn't this lead into another state? This won't tokenise the characters that presumably come after the caret. (Oh, and example in the visual sample, please.)

Reply:
Yes, the caret indicates that the next character is escaped, and if no character indicates continuation of the instruction in the next line. Too complex for me implement that state. I removed the caret.

Reply:
I don't know if you want to try mixing in the `:escape` state to the `:data` state in the same way that I did for the string states or if you're happy to leave this out for now.

Reply:
I'm not sure if that is neccesary. Is needed ?

Reply:
I wasn't sure if escaping is something that is possible 'outside' strings as it were. Usually, I'd say it wouldn't be but in some languages it can (e.g. lisps).

Reply:
Escape a character is possible inside and outside strings.

Reply:
OK. I've mixed in the `:escape` state to the `:data` state so that it can be used outside strings. I think this is now good to go.